### PR TITLE
Add implicit conversion from IArray to immutable.ArraySeq

### PR DESCRIPTION
**Pull request by nicolasstucki**
Based on #11320
Part of #11298

### CODE REVIEW

File: library/src/scala/IArray.scala [hunk 1]
```diff
@@ -273,6 +273,51 @@ object opaques:
     extension [T](arr: IArray[T]) def zip[U: ClassTag](that: IArray[U]): IArray[(T, U)] =
       genericArrayOps(arr).zip(that)
   }
+
+  /** Conversion from IArray to immutable.ArraySeq */
+  given genericWrapIArray[T]: Conversion[IArray[T], scala.collection.immutable.ArraySeq[T]] =
```
Comment (nicolasstucki) [inline library/src/scala/IArray.scala:278]: Based on https://github.com/scala/scala/blob/2.13.x/src/library/scala/Predef.scala#L544

Not sure if this is the best location for this conversion. We could also move it to Predef.

Comment (nicolasstucki) [reply library/src/scala/IArray.scala:278]: Strange, it seems that implicit conversions have higher priority than extension methods. @odersky, is that the intended behaviour?

Comment (nicolasstucki) [reply library/src/scala/IArray.scala]: I had to add a DummyImplicit to make it work

**Review by odersky — CHANGES_REQUESTED** [review]
Should there be a conversion from `IArray` to `Seq`? I see three alternatives:

 1. No conversion, but add all `Seq` methods (including `toSeq`) as extension methods to `IArray`. The methods can be overloaded to specialize for primitive element types, just like the wrapArray conversions are specialized.
 2. An old style implicit conversion
 3. A new style `Conversion` instance.

I think I would go with (1). It's better not to introduce implicit conversions for new types. (2) is better than (3) since new style conversions warn at the use site which is definitely not what we want here. So new style conversions should not be used for extension methods.

I experimented with the idea of overloaded `toSeq`. It seems to be possible to do it if we use `@targetName`:
```scala
import annotation.targetName
case class IA[+X](xs: X*)

extension (x: IA[Int])
  @targetName("toIntSeq")
  def toSeqq: Seq[Int] = List(1, 2, 3, 4)
extension [T](x: IA[T])
  def toSeqq: Seq[T] = List()

def f[T](as: IA[T]) =
  val xs = IA(1, 2, 3)
  println(xs.toSeqq)
  println(as.toSeqq)

@main def Test =
  f(IA("a"))
```
Adding all the `Seq` methods as extension methods is a bit tedious. I have a branch that is already half way there. 

I have another idea how we can avoid the repetition in the future: Allow `export` in extension blocks. But that is for after 3.0.

**Comment by sjrd** [comment]
Why would we specialize all the extension methods of `IArray` when we don't do that for `Array`?

**Comment by odersky** [comment]
> Why would we specialize all the extension methods of IArray when we don't do that for Array?

We only specialize `toSeq` which corresponds to the specialized implicit conversions from `Array` to `ArraySeq`. I.e. the `Seq` wrapper should not have to use reflection to access the array elements.

**Comment by nicolasstucki** [comment]
> I experimented with the idea of overloaded toSeq. It seems to be possible to do it if we use @targetName:

We do not need to use @targetName as the extension methods have the arrays in their signature and those be resolved as they are just arrays for the JVM. We already do that with `apply` and `length`.

**Pull request closed without merging**